---
title: "ATACseq In Bioconductor <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_ATACseq/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# ATACseq (part 2)

---
"    
  )
  
}

```

```{r setup, include=FALSE}
library(soGGi)
library(ChIPQC)
library(GenomicAlignments)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(DESeq2)
library(BSgenome.Mmusculus.UCSC.mm10)
library(tracktables)
library(clusterProfiler)
library(org.Mm.eg.db)

library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)

```

## Data

In this section we will work a little more with the Greenleaf dataset. 

To remind you, this is the original ATACseq dataset and we are working with a single sample. More details are [here.](https://rockefelleruniversity.github.io/RU_ATACseq/presentations/slides/RU_ATAC.html#5)

We will characterize the Greenleaf data signal around TSS regions and identify and annotate open regions using peak calls with MACS2.

At the end of the session we will plot transposase cut-sites around DNA binding proteins.

---

## Data

In this session we will also start working with a larger dataset from the ENCODE consortium generated by Bing Ren at UCSD. This dataset has multiple samples for multiple tissues (Liver, Kidney and Hindbrain), allowing us to perform a differential analysis on open regions.

Links to data and sample information are included in list below.

* Liver day 12 - [ENCSR302LIV](https://www.encodeproject.org/experiments/ENCSR302LIV/)

* Kidney day 15 - [ENCSR023QZX](https://www.encodeproject.org/experiments/ENCSR023QZX/)

* Hindbrain day 12 - [ENCSR088UYE](https://www.encodeproject.org/experiments/ENCSR088UYE/)


---
## Processed Data

We start with public sequencing data in links below and use reference data in Bioconductor.

Since some of these processing steps may take a little time I provide links to pre-processed results.

**Essentials**

BAM file and BAI index from our alignment/sorting/indexing.

* [SAMN02192806 - Greenleaf BAM](https://s3.amazonaws.com/rubioinformatics/ATAC_Workshop/ATAC_Data/ATAC_BAM/Sorted_ATAC_50K_2.bam) - Full BAM file for Greenleaf example produced following in our Rsubread alignment, sorting and indexing.  

* [SAMN02192806 - Greenleaf BAI index](https://s3.amazonaws.com/rubioinformatics/ATAC_Workshop/ATAC_Data/ATAC_BAM/Sorted_ATAC_50K_2.bam.bai) - BAI index file for BAM in Greenleaf example produced following in our alignment, sorting and indexing. 

---
**Essentials**

Small BAM, peak calls and directory structure.

* [ATAC_Workshop_Essential.zip](https://s3.amazonaws.com/rubioinformatics/ATAC_Workshop_Essential.zip) - Require additional workshop files and directory structure.

Once you have downloaded the above and unzipped **ATAC_Workshop.zip**, you should move the **Sorted_ATAC_50K_2.bam** and **Sorted_ATAC_50K_2.bam.bai** file into **ATAC_Workshop/ATAC_Data/ATAC_BAM/**

---
**Not essential**

Same as above but with BAMs for counting as well as small BAM, peak calls and directory structure.

* [ATAC_Workshop.zip](https://s3.amazonaws.com/rubioinformatics/ATAC_Workshop.zip)  - Additional workshop files and directory structure.

Bigwigs for IGV.

* [Bigwigs](https://s3.amazonaws.com/rubioinformatics/ATAC_bigWigs.zip) - BigWigs to review in IGV. 

---
## ATACseq

<div align="center">
<img src="imgs/mnATAC.jpg" alt="offset" height="300" width="600">
</div>

* ATACseq - Uses transposases and offers a method to simultaneously extract signal from transcription factors binding sites and nucleosome positions from a single sample.


---
## Working with ATACseq

This means our data potentially contains a mixture of signal types in our data.

* We will have signal from nucleosome free regions and around transcription factors (our shorter fragments).
* A portion of our signal will be from around nucleosomes in open chromatin (longer fragments).

All our data is from open chromatin where our transposase has been able to access.

<div align="center">
<img src="imgs/buenstro.png" alt="offset" height="300" width="600">
</div>


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Evaluating TSS signal

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Evaluating TSS signal

---
"    
  )
  
}
```


##  Evaluating signal over TSS regions

If our shorter fragments represent the open regions around transcription factors and transcriptional machinery we would expect to see signal at transcriptional start sites.

Our longer fragments will represent signal around nucleosomes and so signal should be outside of the transcriptional start sites and more present at the +1 and -1 nucleosome positions.

<div align="center">
<img src="imgs/phasing.png" alt="offset" height="300" width="350">
</div>

---

##  Evaluating signal over TSS regions

We can create a meta-plot over all TSS regions to illustrate where our nucleosome free and nucleosome occupied fractions of signal are most prevalent.

Meta-plots average or sum signal over sets of regions to identify trends in data.

<div align="center">
<img src="imgs/signalOverTSS.png" alt="offset" height="300" width="600">
</div>

---

##  Plotting signal over regions in R

To produce meta-plots of signal over regions we can use the **soGGi** bioconductor package. We can load in **soGGi** with the BiocManager::install and library function, as we have done before.

```{r processData_aligna, echo=TRUE,eval=FALSE,cache=FALSE}
BiocManager::install("soGGi")
library(soGGi)
```
---

##  Plotting regions in soGGi

The soGGi library simply requires a BAM file and a GRanges of regions over which to average signal to produce the meta-plot.

We wish to plot over TSS regions and so we first will need to produce a GRanges of TSS locations for hg19 genome.

Thankfully we now know how to extract these regions for all genes using the **TxDB packages** and some **GenomicRanges** functions.

First we can load our TxDb of interest - **TxDb.Hsapiens.UCSC.hg19.knownGene**.

```{r processData_txdb, echo=TRUE,eval=TRUE,cache=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
TxDb.Hsapiens.UCSC.hg19.knownGene
```

---

##  Plotting regions in soGGi

We can extract gene locations (TSS to TTS) [using the **genes()** function and our **TxDb** object.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicFeatures_In_Bioconductor.html#15)

```{r processData_genes, echo=TRUE,eval=TRUE,cache=FALSE}
genesLocations <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
genesLocations
```


---
## Plotting regions in soGGi

We can then use the [**resize()** function](https://rockefelleruniversity.github.io/Bioconductor_Introduction/presentations/slides/GenomicIntervals_In_Bioconductor.html#40) to extract the location of start of every gene (the TSSs) in a stranded manner.

Here we set the **fix** position as the start and the width as 1.

```{r processData_resize, echo=TRUE,eval=TRUE,cache=FALSE}
tssLocations <- resize(genesLocations,fix="start",width = 1)
tssLocations

```

---
## Plotting regions in soGGi

When we created our index we subset the genome to the main chromosomes. We can do this again with our TSS GRange object, and update the levels. This means the BAM and GRanges will play nicely. 

```{r processData_subset, echo=TRUE,eval=F}
mainChromosomes <- paste0("chr",c(1:21,"X","Y","M"))

myindex<-(match(seqnames(tssLocations), mainChromosomes))


tssLocations <- tssLocations[as.numeric(myindex)]

seqlevels(tssLocations) <- mainChromosomes

```


```{r processData_subset_fix, echo=F,eval=F}
mainChromosomes <- paste0("chr",c(1:21,"X","Y","M"))

myindex<-(match(seqnames(tssLocations), mainChromosomes))


tssLocations <- tssLocations[!is.na(myindex)]

seqlevels(tssLocations) <- mainChromosomes
save(tssLocations, file="data/tss_pres2.Rmd")

```

---
##  Plotting regions in soGGi

The soGGi package's **regionPlot()** function requires a BAM file of data to plot supplied to **bamFile** parameter and a GRanges to plot over supplied to **testRanges** argument.


```{r, echo=F, eval=T}
load("data/tss_pres2.Rmd")
```

```{r processData_soggi, echo=TRUE,eval=FALSE,cache=FALSE}
library(soGGi)
sortedBAM <- "~/Downloads/ATAC_Workshop/Sorted_ATAC_50K_2.bam"

library(Rsamtools)
# Nucleosome free
allSignal <- regionPlot(bamFile = sortedBAM,
                        testRanges = tssLocations)
```

---
## Plotting regions in soGGi

Additionally we supply information on input file format to **format** parameter, whether data is paired to **paired** parameter and type of plot to **style** parameter. We explore visualization options [visualization training](https://rockefelleruniversity.github.io/RU_VisualizingGenomicsData.html).

A useful feature is that we can we can specify the minimum and maximum fragment lengths of paired reads to be used in our plot with the **minFragmentLength** and **maxFragmentLength** parameters. This allows us to select only our nucleosome free signal (< 100 base-pairs) to produce our metaplot over TSS regions.

```{r processData_soggia1, echo=F,eval=T}
load("data/nucFree_TSS.Rdata")
```

```{r processData_soggia, echo=TRUE,eval=FALSE,cache=FALSE}
nucFree <- regionPlot(bamFile = sortedBAM,
                        testRanges = tssLocations,
                        style = "point",
                        format="bam",
                        paired=TRUE,
                        minFragmentLength = 0,
                        maxFragmentLength = 100,
                        forceFragment = 50)
class(nucFree)
```
---
##  Plotting regions in soGGi

Now we have our profile object we can create our metaplot using the **plotRegion()** function in **soGGi**.

Here we see the expected peak of signal for our nucleosome free region in the region over the TSS.

```{r processData_plot,fig.height=3.5,fig.width=7, echo=TRUE,eval=T,cache=TRUE,message=FALSE,warning=FALSE}
plotRegion(nucFree)
```


---

##  Plotting regions in soGGi

We can create a plot for our mono-nucleosome signal by adjusting our **minFragmentLength** and **maxFragmentLength** parameters to those expected for nucleosome length fragments (here 180 to 240).

```{r processData_soggi4, echo=F,eval=FALSE,cache=FALSE,message=FALSE,warning=FALSE}
monoNuc <- regionPlot(bamFile = sortedBAM,
                        testRanges = tssLocations,
                        style = "point",
                        format="bam",
                        paired=TRUE,
                        minFragmentLength = 180,maxFragmentLength = 240,forceFragment = 80)
save(monoNuc,file = "data/monoNuc_TSS.RData")
```

```{r processData_soggi2.5, echo=F,eval=T,cache=FALSE,message=FALSE,warning=FALSE}

load(file = "data/monoNuc_TSS.RData")
```

```{r processData_soggi3, echo=TRUE,eval=FALSE,cache=FALSE,message=FALSE,warning=FALSE}
monoNuc <- regionPlot(bamFile = sortedBAM,
                        testRanges = tssLocations,
                        style = "point",
                        format="bam",
                        paired=TRUE,
                        minFragmentLength = 180,maxFragmentLength = 240,forceFragment = 80)

```

---
##  Plotting regions in soGGi

Similarly we can plot the mono-nucleosome signal over TSS locations using **plotRegion()** function.

In this plot we can clearly see the expected +1 nucleosome signal peak as well as several other nucleosome signalpeaks

```{r processData_plot3, fig.height=3.5,fig.width=7, echo=TRUE,eval=T,cache=FALSE, message=FALSE, warning=FALSE}
plotRegion(monoNuc)
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Peak calling

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Peak calling

---
"    
  )
  
}
```

## Finding Open Regions

A common goal in ATACseq is to identify nucleosome free regions where transcription factors are binding and/or transcriptional machinery is active. This nucleosome free signal would correspond to fragments less than one nucleosome (as defined in Greenleaf paper < 100bp)

To identify open chromatin however we could simply use all reads which have been properly paired in sequencing (< 2000bp).

For the remainder of the workshop we will look at analysing the nucleosome free portions of the ATACseq data.

---
## Peak calling for nucleosome free regions

There are many methods available to call nucleosome free regions from ATACseq data with many borrowed from ChIPseq analysis.

One very popular and standard peak caller for ATACseq is MAC2.

MACS2 is well established for identifying punctate peaks in ChIPseq data from transcription factors. Though there is no R wrapper, we can easily run MACS2 on the command line using files we create in R, or run it through R using the **system()** function. We cover its installation and use for ChIPseq [here](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor2.html#37) 

---
## Single-end peak calling

With single-end sequencing from ATACseq we do not know how long the fragments are.

To identify open regions therefore requires some different parameters for MACS2 peak calling.

One strategy employed is to shift read 5' ends by -100 and then extend from this by 200bp. Considering the expected size of our nucleosome free fragments this should provide a pile-up over nucelosome regions suitable for MACS2 window size. 

*NOTE: These are command line calls. Wrap these in the __system()__ function if you want to run them through R*
```{r MACS2_singlePeakCallOpen, engine="bash", echo=TRUE,eval=FALSE,cache=TRUE}
MACS2 callpeak -t singleEnd.bam --nomodel --shift -100
                --extsize 200 --format BAM -g MyGenome
```

---
## Single-end peak calling

Alternatively for the nucleosome occupied data we can adjust shift and extension to centre the signal on nucleosome centres (nucleosomes wrapped in 147bp of DNA).  

*NOTE: These are command line calls. Wrap these in the __system()__ function if you want to run them through R*

```{r MACS2_singlePeakCallmononuc, engine="bash", echo=TRUE,eval=FALSE,cache=TRUE}
MACS2 callpeak -t singleEnd.bam --nomodel --shift 37
               --extsize 73 --format BAM -g MyGenome
```
---
## Paired-end peak calling

If we have sequenced paired-end data then we do know the fragment lengths and can provide BAM files to MACS2 which have been prefiltered to properly paired (and fragment size if you want to distinguish nucleosomes from nucleosome free regions)

We have to simply tell MACS2 that the data is paired using the format argument.

This is important as MACS2 will guess it is single-end BAM by default.

*NOTE: These are command line calls. Wrap these in the __system()__ function if you want to run them through R*

```{r MACS2_pairedPeakCall, engine="bash", echo=TRUE,eval=FALSE,cache=TRUE}
MACS2 callpeak -t pairedEnd.bam -f BAMPE 
               --outdir path/to/output/
               --name pairedEndPeakName -g MyGenome
```
---
## Paired-end peak calling
For our paired-end data here, we call peaks on our nucleosome free regions from our nucleosome free region BAM file.

*NOTE: These are command line calls. Wrap these in the __system()__ function if you want to run them through R*

```{r MACS2_pairedPeakCall_Actual, engine="bash", echo=TRUE,eval=FALSE,cache=TRUE,tidy=FALSE,include=TRUE}
MACS2 callpeak  -t ~/Downloads/Sorted_ATAC_50K_2_openRegions.bam
                --outdir ATAC_Data/ATAC_Peaks/ATAC_50K_2
                --name Sorted_ATAC_50K_2_Small_Paired_peaks.narrowPeak
                -f BAMPE -g hs
```

---
## Paired-end peak calling

Following peak calling we would get 3 files.

* Name.narrowPeak -- Narrow peak format suitable for IGV and further analysis

* Name_peaks.xls -- Peak table suitable for review in excel.(not actually a xls but a tsv)

* summits.bed -- Summit positions for peaks useful for finding motifs and plotting

---
## QC

Often we want to do some QC to check for low quality, duplicates and signal distribution.

Before we remove any data we can get a quick assessment of our reads in peaks, duplication rate, low quality reads and reads in artifact regions from ChIPQC.

To do this we will use the hg19 [blacklist](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor2.html#24) which you can download [here](https://www.encodeproject.org/annotations/ENCSR636HFF/)

```{r processData_callQC,messages=FALSE,warning=FALSE, echo=TRUE,eval=FALSE,cache=TRUE,dependson="readinPeakCalling"}
library(ChIPQC)
library(rtracklayer)
library(DT)
library(dplyr)
library(tidyr)

blkList <- import.bed("~/Downloads/ENCFF001TDO.bed.gz")
openRegionPeaks <- "~/Downloads/ATAC_Workshop/ATAC_Data/ATAC_Peaks/Sorted_ATAC_50K_2_Small_Paired_peaks.narrowPeak"

qcRes <- ChIPQCsample("~/Downloads/ATAC_Workshop/ATAC_Data/ATAC_BAM/Sorted_ATAC_50K_2_openRegions.bam",
                      peaks = openRegionPeaks,
                      annotation ="hg19",
                      chromosomes = "chr20",
                      blacklist = blkList,
                      verboseT = FALSE)
```

---
## QC

We can use the ChIPQC package to capture some important metrics for our ATACseq data such as reads in peaks and reads in blacklist from the **QCmetrics()** function and numbers of duplicated reads from the **flagtagcounts()** funcion.

```{r ded, include=FALSE,cache=TRUE}
library(ChIPQC)
library(rtracklayer)
load("data/qcRes.RData")

blkList <- import.bed("data/ENCFF001TDO.bed.gz")
openRegionPeaks <- "data/Sorted_ATAC_50K_2_Small_Paired_peaks.narrowPeak"

```

```{r processData_callQC2,messages=FALSE,warning=FALSE, echo=TRUE,eval=TRUE,cache=TRUE,dependson="ded"}
myMetrics <- QCmetrics(qcRes)
myMetrics[c("RiBL%","RiP%")]
flgCounts <- flagtagcounts(qcRes)
DupRate <- flgCounts["DuplicateByChIPQC"]/flgCounts["Mapped"]
DupRate*100
```

---
##  Remove blacklisted peaks 

Artifacts from sequencing and imperfect genome builds can confound our results. These artifacts have been collated together into "Blacklists" of regions. We discuss these regions in more detail [here](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor2.html#24)

Since blacklisted regions may confound our analysis we remove any peaks which have been called there. 

Removing blacklists too early can hide some of the QC issues in your data. The **blacklist** should always be considered in your analysis and recommended to removed data from these regions once QC is considered.

```{r processData_filterBLKlist, echo=F,eval=F,cache=TRUE,dependson="processData_callQC"}

MacsCalls <- granges(qcRes[seqnames(qcRes) %in% "chr20"])

data.frame(Blacklisted=sum(MacsCalls %over% blkList),
           Not_Blacklisted=sum(!MacsCalls %over% blkList))
save(MacsCalls,file="data/MacsCall_pres2.RData")

```

```{r , echo=T,eval=F}

MacsCalls <- granges(qcRes[seqnames(qcRes) %in% "chr20"])

data.frame(Blacklisted=sum(MacsCalls %over% blkList),
           Not_Blacklisted=sum(!MacsCalls %over% blkList))

MacsCalls <- MacsCalls[!MacsCalls %over% blkList]

```

```{r, echo=F, eval=T}
load("data/MacsCall_pres2.RData")
MacsCalls <- MacsCalls[!MacsCalls %over% blkList]

```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Annotating Peaks

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Annotating Peaks

---
"    
  )
  
}
```

## Annotating Open Regions

It is often of interest to associate identified nucleosome free regions to genomic features such as genes and enhancers.

Once annotated to genes or enhancers' genes, we can start to associate ATACseq data to characteristics of these genes. (functional annotation, expression changes, other epigenetic states)

---
## Annotating peaks to genes

A simple approach to annotating nucleosome free region to genes is to associate regions to their closest gene or within a window around a genes transcriptional start site.

We can use the [chipseeker library to identify genes closest to our regions and to give us simple summaries and visualizations of this annotation.](https://rockefelleruniversity.github.io/RU_ChIPseq/chipseq_course/Presentations/Slides/ChIPseq_In_Bioconductor2.html#48)

We use the gene models from TxDb.Hsapiens.UCSC.hg19.knownGene and supply this to ChIPseeker packages annotatePeak function.

ChIPseeker's csAnno object will then show breakdown of percentages of peaks in genomic regions.

```{r processData_annotatePeak, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_filterBLKlist"}
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
MacsCalls_Anno <-  annotatePeak(MacsCalls,
                                TxDb = TxDb.Hsapiens.UCSC.hg19.knownGene)
MacsCalls_Anno
```

---
## Displaying annotation distribution

As well as showing us a table of the annotation distribution we can visualize this using the plotAnnoPie and plotAnnoBar functions.

```{r processData_Pie, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_annotatePeak"}
plotAnnoPie(MacsCalls_Anno)
```

---

## Annotating nucleosome free regions

With this information we can then subset our peaks/nuc free regions to those only landing in TSS regions (+/- 500)

```{r processData_annotated, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_annotatePeak"}
MacsGR_Anno <- as.GRanges(MacsCalls_Anno)
MacsGR_TSS <-   MacsGR_Anno[abs(MacsGR_Anno$distanceToTSS) < 500]
MacsGR_TSS[1,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Functional analysis of peaks

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Functional analysis of peaks

---
"    
  )
  
}
```

## Functional Analysis of NF regions

Another common step to ATACseq analysis is to identify any functional enrichment in genes associated to nucleosome free regions.

One approach is to take the genes we identified from ChIPseeker as having nuclesome free regions and test these for functional enrichment using [standard tools like GOseq](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor3.html#7).

[Another approach which is well suited to ATACseq is that implemented in GREAT.](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor3.html#17)

**rGREAT by default will limit our queries and so we may need to be patient. This can be adjusted in rGREAT options.**

---
## Functional Analysis 

We can submit our peak calls to GREAT using the **submitGreatJob** function and review available categories of results using **availableCategories**.

```{r processData_funAnalysise,echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_filterBLKlist"}
library(rGREAT)
great_Job <- submitGreatJob(MacsCalls, species = "hg19")
availableCategories(great_Job)
```
---
## Functional Analysis

For this example we select the results tables for the GO category using getEnrichmentTables functions and then review the results for Biological processes.

```{r processData_funAnalysis2,echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_funAnalysis2",message=FALSE, warning=FALSE}
great_ResultTable = getEnrichmentTables(great_Job, category = "GO")
names(great_ResultTable)          
great_ResultTable[["GO Biological Process"]][1:4, ]

```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Differential ATACseq

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Differential ATACseq

---
"    
  )
  
}
```


## Differential ATACseq

We have briefly reviewed the processing and initial analysis of one ATACseq sample using R.

In the next part we will look at how we can identify changes in open regions using R/Bioconductor.

Here we will take an approach akin that in Diffbind and reasonably esatablished in ATACseq analysis. 

---
## Identifying non-redundant peaks

First, We will define a set of non-redundant peaks present in at least 2 samples and use these to assess changes in nucleosome free ATACseq signal using DESeq2. [Here we use the same method for deriving consensus peaks for differentials as seen for ChIPseq.](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor4.html#22)

We take peaks across all our samples and reduce them into a single set of non-redudnant peaks. We can then create a matrix of presence/absence of these peaks over each sample.  


```{r processData_consensus_read, echo=F,eval=T}
peaks <- dir("data/ATAC_Peaks_forCounting/",
             pattern="*.narrowPeak",full.names=TRUE)
```

```{r processData_consensusa, echo=TRUE,eval=F}
peaks <- dir("~/Downloads/ATAC_Workshop/ATAC_Data/ATAC_Peaks_forCounting/",
             pattern="*.narrowPeak",full.names=TRUE)
```

```{r processData_consensus_read2, echo=TRUE,eval=T}
myPeaks <- lapply(peaks,ChIPQC:::GetGRanges,simple=TRUE)
allPeaksSet_nR <- reduce(unlist(GRangesList(myPeaks)))
overlap <- list()
for(i in 1:length(myPeaks)){
  overlap[[i]] <- allPeaksSet_nR %over% myPeaks[[i]]
}
overlapMatrix <- do.call(cbind,overlap)
colnames(overlapMatrix) <- basename(peaks)
mcols(allPeaksSet_nR) <- overlapMatrix

```

---
## Identifying non-redundant peaks


```{r processData_consensusaa, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_consensusa"}

allPeaksSet_nR[1:2,]
```
---
## Identifying non-redundant peaks

We filter out peaks in blacklists and in ChrM prior to testing to eliminate potential artifact differential calls.

```{r processData_consensusCounting, echo=F,eval=F,cache=TRUE,dependson="processData_consensusa"}
blklist <- import.bed("data/ENCFF547MET.bed.gz")
nrToCount <- allPeaksSet_nR[!allPeaksSet_nR %over% blklist 
                            & !seqnames(allPeaksSet_nR) %in% "chrM"]
save(nrToCount,file="data/nrToCount_pres2.RData")
```

```{r, eval=F, echo=T}
blklist <- import.bed("data/ENCFF547MET.bed.gz")
nrToCount <- allPeaksSet_nR[!allPeaksSet_nR %over% blklist 
                            & !seqnames(allPeaksSet_nR) %in% "chrM"]
nrToCount
```

```{r, eval=T, echo=F}
load("data/nrToCount_pres2.RData")
nrToCount
```

---
## Counting for differential ATACseq

We now identify the number of samples in which our non-redundant peaks occur. Here we use the **rowSums()** function with our occurrence matrix and select those samples occuring in at least 2 samples.

```{r processData_consensusCountingf, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_consensus"}
library(Rsubread)
occurrences <- rowSums(
  as.data.frame(elementMetadata(nrToCount)
                )
  )

nrToCount <- nrToCount[occurrences >= 2,]
nrToCount
```
---
## Counting for differential ATACseq

Now we have to set of regions to count in we can use **summariseOverlaps()** to count paired reads landing in peaks as we have done for ChIPseq. 

We have to adjust our counting for paired-end reads so we additionally set the **singleEnd** parameter to FALSE.


```{r processData_consensusCounting2, echo=TRUE,eval=FALSE,cache=TRUE,dependson="processData_consensusCountingf"}
library(GenomicAlignments)
bamsToCount <- dir("~/Downloads/ATAC_Workshop/ATAC_Data/ATAC_BAM_forCounting/",
                   full.names = TRUE,pattern = "*.\\.bam$")

myCounts <- summarizeOverlaps(consensusToCount,
                              bamsToCount,singleEnd=FALSE)

colnames(myCounts) <- c("HindBrain_1","HindBrain_2","Kidney_1","Kidney_2",
                        "Liver_1","Liver_2")
```
---
## DESeq2 for differential ATACseq

With our counts of fragments in nucleosome free regions we can now contruct a DESeq2 object.

We pass the GRanges of regions we count to DESeqDataSetFromMatrix function so as to access these from DESeq2 later.



```{r processData_DEseq2_PCA, echo=TRUE,eval=TRUE,cache=TRUE}
library(DESeq2)
load("data/myCounts.RData")
Group <- factor(c("HindBrain","HindBrain","Kidney","Kidney",
                  "Liver","Liver"))
metaData <- data.frame(Group,row.names=colnames(myCounts))

atacDDS <- DESeqDataSetFromMatrix(assay(myCounts),
                                  metaData,
                                  ~Group,
                                  rowRanges=rowRanges(myCounts))
atacDDS <- DESeq(atacDDS)

```
---
## DESeq2 for differential ATACseq

With the new DESeq2 object we can now test for any differences in ATACseq signal between groups.

In this example we look at differences between the Hindbrain and Kidney samples. 

We return a GRanges object here to allow us to perform some more GenomicRanges operations.

```{r processData_DEseq2_Results_ResultsTable, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_DEseq2_PCA"}
KidneyMinusHindbrain <- results(atacDDS,
                                c("Group","Kidney","HindBrain"),
                               format="GRanges")
KidneyMinusHindbrain <- KidneyMinusHindbrain[
                                order(KidneyMinusHindbrain$pvalue)
                         ]
KidneyMinusHindbrain
```

---
## DESeq2 for differential ATACseq

We can subset to only open regions within promoters and then create a HTML table to review the results in IGV using the **makebedtable()** function in tracktables package. 

You can see the result [here](https://rockefelleruniversity.github.io/RU_ATACseq/data/KidneyMinusHindbrain.html)

```{r processData_DEseq2_ResultsToTSSregions,message=FALSE,warning=FALSE, echo=F,eval=TRUE,cache=TRUE,dependson="processData_DEseq2_Results_ResultsTable"}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
toOverLap <- promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,
                       500,500)
KidneyMinusHindbrain <- KidneyMinusHindbrain[
                          (!is.na(KidneyMinusHindbrain$padj) 
                           & KidneyMinusHindbrain$padj < 0.05) 
                           & KidneyMinusHindbrain %over% toOverLap,]
makebedtable(KidneyMinusHindbrain,"KidneyMinusHindbrain.html",'data/')
```

```{r processData_DEseq2_ResultsToTSSregions2,message=FALSE,warning=FALSE, echo=TRUE,eval=F, cache=TRUE,dependson="processData_DEseq2_Results_ResultsTable"}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
toOverLap <- promoters(TxDb.Mmusculus.UCSC.mm10.knownGene,
                       500,500)
KidneyMinusHindbrain <- KidneyMinusHindbrain[
                          (!is.na(KidneyMinusHindbrain$padj) 
                           & KidneyMinusHindbrain$padj < 0.05) 
                           & KidneyMinusHindbrain %over% toOverLap,]
myReport<-makebedtable(KidneyMinusHindbrain,"KidneyMinusHindbrain.html",getwd())
browseURL(myReport)
```

---
## Annotation for differential ATACseq

In the final part we can annotate our differential ATACseq regions to genes and then use gene information to test enrichment for GO sets.

Since we have subset regions to those within +/- 500bp of a TSS we can use a standard enrichment analysis at this point. Here we use clusterProfiler to identify enrichment

```{r processData_DEseq2_functionalEnrichmentAnalysiss, echo=TRUE,eval=TRUE,cache=TRUE, dependson="processData_DEseq2_ResultsToTSSregions",message=FALSE,warning=FALSE}

anno_KidneyMinusHindbrain <- annotatePeak(KidneyMinusHindbrain,
                                          TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,verbose = FALSE)
DB_ATAC <- as.data.frame(anno_KidneyMinusHindbrain)
DB_ATAC[1,]
```

---
## Annotation for differential ATACseq

Since we have subset regions to those within +/- 500bp of a TSS we can use a standard enrichment analysis at this point. Here we use clusterProfiler to identify enrichment. For alternatives have a look at our [ChIPseq course](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor3.html#7)

```{r processData_DEseq2_functionalEnrichmentAnalysisd, echo=TRUE,eval=TRUE,cache=TRUE, dependson="processData_DEseq2_functionalEnrichmentAnalysiss",message=FALSE,warning=FALSE}
library(clusterProfiler)
go <- enrichGO(DB_ATAC$geneId, 
                OrgDb = "org.Mm.eg.db",ont = "BP",maxGSSize = 5000)
go[1:2,1:6]

```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Finding Motifs in ATACseq data

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Finding Motifs in ATACseq data

---
"    
  )
  
}
```

## Cutting sites from ATACseq data

ATACseq should generate shorter fragments (our nucleosome free regions) around smaller *protected* areas such as transcription factor binding sites.

We can therefore look for the pile-up of cut-sites around motifs of interest within different tissues/celltypes/samples.

To produce cut-sites from our BAM file we first resize our reads to 1bp and make the shift of 4/-5 bp depending on strand to adjust for expected shift from insertion of Tn5 transposase.

Here we will identify CTCF motifs passing an arbitary cut-off and then use soGGi to plot cut-sites around them. We will jump back to our Greenleaf dataset to do this.

---
## Finding motifs

We need to identify the position of CTCF motifs across the genome so first we need to know what a CTCF motif looks like.

The motifDB package contains information on Motifs from public databases such as JASPAR. Here we use the **query()** function with our motif of interest (**CTCF**) to extract the CTCF motif.

```{r processData_motifCutsa, echo=TRUE,eval=TRUE,cache=TRUE,message=FALSE,warning=FALSE}
library(MotifDb)
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg19)
CTCF <- query(MotifDb, c("CTCF"))
CTCF
```
---
## Finding motifs

We can extract a point weight matrix for CTCF which specifies the likelihood of a DNA base occurring in a CTCF motif. Here we extract the motif for CTCF from Human JASPAR Core database.

```{r processData_motifCutsb, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsa"}
names(CTCF)
ctcfMotif <- CTCF[[1]]
ctcfMotif[,1:4]
```
---
## Visualising PWMs

We can visualise the frequency of DNA bases in our motif using the **seqLogo** package and the **seqLogo** function. 

```{r processData_motifCutsc, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsb",fig.height=5,fig.width=7}
library(seqLogo)
seqLogo(ctcfMotif)
```
---
## Searching for PWMs in DNAstring

We can now use the **matchPWM()** function with our newly acquired CTCF PWM. 

Here we will search the sequence on Chr20 using the sequence provided within the BSgenome library for human **BSgenome.Hsapiens.UCSC.hg19**. 

The result is a **Views** object, similar to the IRanges object. We convert

```{r processData_motifCutsd, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsc",warning=FALSE,message=FALSE}

myRes <- matchPWM(ctcfMotif,BSgenome.Hsapiens.UCSC.hg19[["chr20"]])
myRes
```

---
  
## Searching for PWMs in DNAstring

We need to convert the Views object to a GRanges so we can use these in soGGi to plot cut sites over.

```{r processData_motifCutse, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsd"}

toCompare <- GRanges("chr20",ranges(myRes))
toCompare
```

---
## Shifting reads for cut-sites

To plot cut-sites we will wish to consider only the 5' end of reads and will need to adjust for a known offset of 5' reads to actual T5 cut-sites.

This will involve capturing the 5'end of reads and shifting reads on positive and negative strand by 4bp or -5bp respectively.

First we read in our nucleosome free region BAM file and extract read pairs.



```{r processData_motifCutsdwdw, echo=TRUE,eval=FALSE}
BAM <- "~/Downloads/ATAC_Workshop/ATAC_Data/ATAC_BAM/Sorted_ATAC_50K_2_openRegions.bam"
atacReads_Open <- readGAlignmentPairs(BAM)
read1 <- first(atacReads_Open)
read2 <- second(atacReads_Open)
read2[1,]
```

```{r, echo=F, eval=F}
save(read1, read2, file="data/pres2_reads1and2.RData")


```

```{r processData_motifCutsfa, echo=FALSE,eval=TRUE}
load(file="data/pres2_reads1and2.RData")
read2[1,]
```

---
  
## Shifting reads for cut-sites

Now we can shift the 5' end of both reads pairs by 4bp or -5bp depending on strand. This produces a GRanges of all our cut-sites from both reads.

```{r processData_motifCutsf, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsfa"}

Firsts <- resize(granges(read1),fix="start",1)
First_Pos_toCut <- shift(granges(Firsts[strand(read1) == "+"]),
                                         4)
First_Neg_toCut <- shift(granges(Firsts[strand(read1) == "-"]),
                                         -5)

Seconds <- resize(granges(read2),fix="start",1)
Second_Pos_toCut <- shift(granges(Seconds[strand(read2) == "+"]),
                                4)
Second_Neg_toCut <- shift(granges(Seconds[strand(read2) == "-"]),
                                -5)

test_toCut <- c(First_Pos_toCut,First_Neg_toCut,
                Second_Pos_toCut,Second_Neg_toCut)
test_toCut[1:2,]
```
---
  
## Coverage for cut-sites

Now we can use the GRanges of cut-site positions to produce an RLElist of cut-sites across the genome using the **coverage()** function.


```{r processData_motifCutsg, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsf"}

cutsCoverage <- coverage(test_toCut)
cutsCoverage20 <- cutsCoverage["chr20"]
cutsCoverage20[[1]]
```
---
  
## Plotting for cut-sites

We can use an RLElist with soGGi to produce a plot of cut-sites around our discovered CTCF motifs.

We change the **format** to rlelist and the **distanceAround** parameter to 500bp.

```{r processData_motifCutsh, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsg",message=FALSE,warning=FALSE,fig.height=5,fig.width=7}


CTCF_Cuts_open <- regionPlot(cutsCoverage20,
                         testRanges = toCompare,
                         style = "point",
                         format="rlelist",distanceAround = 500)

```

---
  
## Plotting for cut-sites

Now we can now plot our cut-sites using the **plotRegion()** function.

```{r processData_motifCutsi, echo=TRUE,eval=TRUE,cache=TRUE,dependson="processData_motifCutsh",message=FALSE,warning=FALSE,fig.height=5,fig.width=7}

plotRegion(CTCF_Cuts_open,outliers = 0.001)+
  ggtitle("NucFree Cuts Centred on CTCF")+theme_bw()

```

---
## Time for an exercise!

Exercise on ATACseq data can be found [here](../../exercises/exercises/ATACseq_part2_exercise.html)

---
## Answers to exercise

Answers can be found [here](../../exercises/answers/ATACseq_part2_answers.html)



